File: VetoGame.jl
module VetoGame
    # EXPORTS AND IMPORTS
    export solve_veto_game, solve_veto_game_letters
    using Combinatorics: combinations
    using DataStructures: Queue, dequeue!, enqueue!
    
    # MUTABLE STRUCT: `StateNode`
    mutable struct StateNode
        state::Vector{Int64}
        parent::Union{Nothing, StateNode}
        children::Vector{StateNode}
        outcome::Int64
        depth::Int64
    end
    
    # FUNCTION: `solve_veto_game`
    function solve_veto_game(preferences::Matrix{Int64})
        (m, n) = size(preferences)
        
        if m != n + 1
            throw(ArgumentError(
                "`preferences` must have `n` columns (players) and `n + 1` rows (options)."
            ))
        end
        
        root = StateNode(1:m, nothing, StateNode[], 0, 0)
        Q = Queue{StateNode}()
        enqueue!(Q, root)
        
        while first(Q).depth < n
            node = dequeue!(Q)
            depth = node.depth
            
            for comb in combinations(node.state, m - depth - 1)
                child = StateNode(comb, node, StateNode[], 0, depth + 1)
                push!(node.children, child)
                enqueue!(Q, child)
            end
        end
        
        while first(Q).depth == n
            node = dequeue!(Q)
            node.outcome = node.state[1]
            enqueue!(Q, node.parent)
        end
        
        while first(Q).depth > 0
            parent = first(Q).parent
            
            while !isempty(Q) && first(Q).parent == parent
                node = dequeue!(Q)
                set_outcome!(node, preferences)
            end
            
            enqueue!(Q, node.parent)
        end
        
        set_outcome!(root, preferences)
        solution = Vector{Int64}(undef, n)
        node = root
        
        while node.depth < n
            child = node.children[1]
            solution[node.depth + 1] = setdiff(node.state, child.state)[1]
            node = child
        end
        
        return solution
    end
    
    # FUNCTION: `solve_veto_game_letters`
    # function solve_veto_game_letters(preferences::Matrix{Char}, hash::Dict{Char, Int64})
    #     hash_rev = Dict(values(hash) .=> keys(hash))
        
    #     for (i, k) in enumerate(solve_veto_game([hash[k] for k in preferences]))
    #         println("Player $i vetoes $(hash_rev[k])")
    #     end
    # end
    function solve_veto_game_letters(preferences::Matrix)
        options = unique(preferences)
        hash = Dict(option => i for (i, option) in enumerate(options))
        hash_rev = Dict(values(hash) .=> keys(hash))
        solution = solve_veto_game([hash[k] for k in preferences])
        
        for (i, k) in enumerate(solution)
            println("Player $i vetoes $(hash_rev[k])")
        end
        
        println("\nEnd result: $(setdiff(options, solution)[1])")
    end
    
    # FUNCTION: `set_outcome!`
    function set_outcome!(node::StateNode, preferences::Matrix{Int64})
        player_prefs = @view preferences[:, node.depth + 1]
        sort!(node.children, by = x -> findfirst(player_prefs .== x.outcome))
        
        child = node.children[1]
        node.children = [child]
        node.outcome = child.outcome
        return node
    end
end

File: C++/veto_game.cpp
#include <cstdio> ///
#include <iostream> ///
#include <cstdlib>
#include <algorithm>
#include <queue>
#include "veto_game.hpp"

VetoGameResult solve_veto_game(char **preferences, int n) {
    char *options = preferences[0];
    // std::vector<char> options;
    // options.assign(preferences[0], preferences[0] + n + 1);
    // for (char c : options) { ///
    //     std::cout << c << ','; ///
    // } ///
    // std::cout << std::endl; ///
    
    VetoGameState root = VetoGameState(
        options,
        nullptr,
        std::vector<VetoGameState>(),
        nullptr,
        0,
        0,
        n
    );
    for (int i = 0; i < n + 1; i ++) {
        std::cout << root.state[i] << ','; ///
    }
    printf("Meow\n"); ///
    
    std::queue<VetoGameState*> Q;
    Q.push(&root);
    VetoGameState *node = &root;
    int depth = 0;
    
    while (depth < n) {
        Q.pop();
        // if (node.depth == 1) { ///
        //     printf("%d | ", node.depth); ///
        //     printf("%d\n", node.parent->depth); ///
        // } ///
        
        for (int i = 0; i < n - depth + 1; i++) {
            // char state_child[n - depth];
            char *state_child = new char[n - depth];
            
            for (int j = 0; j < i; j++) {
                state_child[j] = node->state[j];
            }
            
            for (int j = i; j < n - depth; j++) {
                state_child[j] = node->state[j + 1];
            }
            
            // std::vector<char> state_child;
            
            // for (int j = 0; j < i; j++) {
            //     state_child.push_back(node->state[j]);
            // }
            
            // for (int j = i + 1; j < n - depth + 1; j++) {
            //     state_child.push_back(node->state[j]);
            // }
            
            VetoGameState child = VetoGameState(
                // std::vector<char>(),
                // &state_child[0],
                state_child,
                node,
                std::vector<VetoGameState>(),
                nullptr,
                depth + 1,
                0,
                n
            );
            
            // for (int j = 0; j < i; j++) {
            //     child.state.push_back(node->state[j]);
            // }
            
            // for (int j = i + 1; j < n - depth + 1; j++) {
            //     child.state.push_back(node->state[j]);
            // }
            
            node->children.push_back(child); /// Not working here?
            node->num_children++;
            printf("%ul\n", node->children.size()); /// Error here...
            printf("%d\n", node->num_children); /// Error here...
            Q.push(&child);
            // Q.push(&node->children.back());
            if (child.depth < 4) {
                for (int i = 0; i < n - depth; i++) {
                    std::cout << child.state[i] << ','; ///
                }
                printf("\n"); ///
                printf("%d\n", &state_child); ///
                printf("%d\n", &child); ///
            }
            // if (child.depth < 4) { ///
            //     for (char c : child.state) { ///
            //         std::cout << c << ','; ///
            //     }
            //     std::cout << '|'; ///
            //     for (char c : child.parent->state) { ///
            //         std::cout << c << ','; ///
            //     }
            //     std::cout << '|'; ///
            //     for (char c : Q.back()->state) { ///
            //         std::cout << c << ','; ///
            //     }
            //     std::cout << std::endl; ///
            // } ///
        }
        
        // for (char c : node->children[8].state) { ///
        //     std::cout << c << ','; ///
        // }
        // std::cout << '|'; ///
        // for (char c : Q.front()->state) { ///
        //     std::cout << c << ','; ///
        // }
        // std::cout << std::endl; ///
        // printf("%d,|", &Q.front()); ///
        // printf("%d\n", &node->children[8]); ///
        
        // std::cout << std::endl; ///
        printf("%d\n", node->num_children); ///
        node = Q.front();
        depth = node->depth;
        printf("%d\n", &node); ///
        printf("%d | ", depth); ///
        for (int i = 0; i < n - depth; i++) {
            std::cout << node->state[i] << ','; ///
        }
        printf("\n"); ///
        // for (char c : node->state) { ///
        //     std::cout << c << ','; ///
        // }
        // std::cout << std::endl; ///
    }
    
    // for (char c : node.state) { ///
    //     std::cout << c << ','; ///
    // }
    // std::cout << std::endl; ///
    // for (char c : node.parent->state) { ///
    //     std::cout << c << ','; ///
    // }
    // VetoGameState m = root.children[0];
    printf("%d\n", root.children.size()); ///
    for (int i = 0; i < n; i ++) {
        std::cout << root.children[0].state[i] << ','; ///
    }
    printf("\n"); ///
    for (int i = 0; i < n; i ++) {
        std::cout << root.children[1].state[i] << ','; ///
    }
    printf("\n"); ///
    for (int i = 0; i < n; i ++) {
        std::cout << root.children[8].state[i] << ','; ///
    }
    printf("\n"); ///
    
    for (int i = 0; i < n; i ++) {
        std::cout << node->state[i] << ','; ///
    }
    printf("Meow\n"); ///
    while (depth == n) {
        Q.pop();
        node->outcome = node->state;
        // node->outcome = node->state.data();
        node->num_children = 1;
        Q.push(node->parent);
        
        node = Q.front();
        depth = node->depth;
        // for (char c : node.state) { ///
        //     std::cout << c << ','; ///
        // }
        // std::cout << std::endl; ///
        // for (char c : node.parent->state) { ///
        //     std::cout << c << ','; ///
        // }
        // std::cout << std::endl; ///
    }
    
    printf("C\n"); ///
    while (depth > 0) {
        Q.pop();
        VetoGameState *parent = node->parent;
        
        while (!Q.empty() && node->parent == parent) {
            Q.pop();
            node->set_outcome(preferences);
            node = Q.front();
        }
        
        depth = node->depth;
        Q.push(node->parent);
    }
    
    root.set_outcome(preferences);
    char *solution = new char[n];
    node = &root;
    depth = 0;
    
    printf("D\n"); ///
    while (depth < n) {
        VetoGameState *child = &node->children[0];
        
        solution[node->depth] = *std::set_difference(
            node->state,
            node->state + n + 1,
            child->state,
            child->state + n + 1,
            new char[1]
        );
        // solution[node->depth] = *std::set_difference(
        //     node->state.begin(),
        //     node->state.end(),
        //     child->state.begin(),
        //     child->state.end(),
        //     new char[1]
        // );
        
        node = child;
        depth = node->depth;
    }
    
    return VetoGameResult(solution, node->outcome[0], n);
}